### ISSUE TITLE
Re: #959  rudimentary install fails

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#998]
with #959 0.3.1-116-g4f82d5d

``` shell
hopa:~
$> datalad search bids 
No DataLad dataset found at current location
Would you like to install stock DataLad meta-dataset under '/home/yoh/datalad'? (choices: yes, no): yes

2016-10-13 00:50:45,099 [INFO   ] Installing dataset at: /home/yoh/datalad 
You can in future refer to that dataset using -d///
2016-10-13 00:50:45,443 [INFO   ] Performing search using central dataset '/home/yoh/datalad' 
2016-10-13 00:50:45,444 [INFO   ] Loading and caching local meta-data... might take a few seconds 
datalad/dicoms/dartmouth-phantoms/bids_test3 
datalad/labs 
datalad/labs/haxby 
datalad/labs/haxby/raiders 
datalad/openfmri 
datalad/openfmri/ds000001 
...

hopa:~
$> datalad install datalad/openfmri/ds000001
2016-10-13 00:50:57,419 [INFO   ] Installing dataset at: /home/yoh/ds000001 
2016-10-13 00:50:57,422 [ERROR  ] 'NoneType' object has no attribute 'split' [utils.py:_get_flexible_url_candidates:270] (AttributeError) 
```

I guess I should have used 'get -n' for that which seems to work, but such a failure, with what was working before is not kosher to say the least (still not sure why install has to be this dump and cant do what it did before :-))

**Comment by mih** [comment]
OK, I really wanted to please you on this one... but after thinking long and hard how to justify this I just I can't....

First, in your context, your desire makes sense. But datalad does not have this kind of process information. We only have

````
datalad install datalad/openfmri/ds000001
````

where according to the docs the argument is the installation _source_. So you are asking to install from a source that does not exist. You want it to assume that the argument is correct, and to try to explore the given path whether it can eventually be resolved into a dataset, by trying to install possible subdatasets along the way (and maybe fail). In your case this would work. But if we want that, we want to revert the essence of #959.

What #959 is about is to increase the amount of evidence in a call that helps datalad increase the probability to make the right predictions. In your case, there is a non-zero chance that the non-existing path is actually a malformed argument (missing `~/` prefix, etc.). but we don't know. With #959 I want people to use `get` for this case, because by doing so they indicate that they want to obtain a dataset component. I believe that this is a better justification to started cloning repositories along the way in order to explore whether an argument is valid.

**Comment by bpoldrack** [comment]
First of all, I think you DID justify it ;-)
And I agree.

Nevertheless, there's a way to also please @yarikoptic , I guess. It boils down the issue, that datalad atm isn't able to recognize, that `datalad/openfmri/ds000001` CAN be treated as a "source" in a way. So the question is, how could we possibly know, that this call makes sense? And we can by doing it the same way Yarik did: Asking the metadata (or even the top-level sibling). Should be possible to integrate such a mechanism into the `RI`classes, so that an instance would return proper source and target for install, no? That way, there's no need to blow up install again.
Does this sound reasonable?

**Comment by yarikoptic** [comment]
Isn't this UC10 in our API doc where you did point that you co do the same install?

**Comment by mih** [comment]
This is a philosophical issue IMHO. Indeed we could ask the metadata whether it knows anything about this subdataset. And I think this should be done in `get`, because we don't want to avoid explorative installation as much as possible.

But in this case, it simply isn't clean. #959 wants `install` to install from a `source`.

The first command in Yarik's original code actually installed the umbrella dataset in `$HOME/datalad`. If the actual scenario is that I simply forgot the `~/` prefix I will end up with a result after datalad wanted to be helpful and do its best to make sense of my argument, but it will not yield the results I wanted. And I actually would have appreciated datalad to say "no fucking way" ;-)

**Comment by yarikoptic** [comment]
I might be still missing the point:  what would be wrong in such a logic (if no explicit --source is provided, typing without trying to recall any of our helpers):

```python
urls_to_install = []
paths_to_get = []
for url_or_path in INPUTS:
   if not is instance(RI(url_or_path), Path):  # so we have some kind of a URL
      urls_to_install.append(url_or_path)  
   else:  # we did get a local path -- just call
      paths_to_get.append(url_or_path)

for url in urls_to_install:
   install(path_from_url(url), source=url, ...)

if paths_to_get:
   get(paths_to_get, get_data=False)

```

?

**Comment by mih** [comment]
We are talking past each other. I am saying #959 does not want a `<figuremeoutIcouldbeanything>` argument. It could be made to work, sure (one way is to simply not merge #959), but the whole point is to **not** have it working.

**Comment by mih** [comment]
Maybe a more cartoonish example:

You want to see this working:

```
datalad install datalad/openfmri/ds000001
```

but why so complicated:

```
datalad datalad/openfmri/ds000001
```

it is only one more deduction to be reasonable sure that you want to `install` this. So why force users into having to use commands, when they could be guessed (p<0.05).

Again, a philosophical issue. And I don't want that. I want the thing to fail, whenever there is reasonable doubt -- which is of course absolutely subjective ;-)

**Comment by yarikoptic** [comment]
and I am trying to avoid philosophical issues and go with a pragmatic approach :-P    Here is some chronology
- we first danced around "install does everything including kitchensink" -- I whined, but agreed to the experiment.  Well -- it didn't work since that is when lots of guesswork was done which broke in many ways
- create was taken out of install ... API became more usable
- add was taken out of install ... API became even more usable
- get was taken out to simplify implementation ... API became very usable, implementation was convoluted (largely due to legacy of this chronology), but there were little to no confusing behaviours at least at the UX level.  Some functionality was not supported/not-implemented but otherwise, as you could see in the API table -- pretty much everything behaves as someone could expect.

In this PR you did great job simplifying implementation, but IMHO it doesn't require crippling `install` at all. The logic I outlined in above snippet is imho very reasonable (unless you give me a counter example) and very easy to document, thus it will not be confusing.  To me it is more confusing why I can't install multiple datasets using `install` when I can do so using `get`, whenever both commands share lots of code/usecases (as you can see from "Needs " columns in API spreadsheet, and almost always having both get and install ways to achieve those use-cases we outlined).  Or why I have to use `get -n` whenever underneath it does 'install' anyways?

**Comment by bpoldrack** [comment]
Well, I'll try to reword the idea, that `install` needs a source, whereas a `get` doesn't. May be that helps:
I think the difference between `get` and `install` here is, that the command itself states, that datalad is supposed to know the thing already in case of `get`, whereas `install` gets things, we don't know anything about yet (may be call "truly remote"). So, the distinction reduces the ambiguity of what a parameter's meaning can possibly be.

This indeed strikes out a lot of special cases and simplifies logic. And therefore it makes sense to have to use `get` here. Yes, it looks kind of the same, but it isn't. By using `get` I tell datalad, that it should be able to derive all needed information from existing datasets (and therefore I want it to try). If I'd use `install` for that, datalad would need to guess and either not fail, when it should or do a lot of (possibly expensive) stuff before failing. The alternative would be a `--IfYouDontKnowTheThingTryToMakeSenseOfIt` switch for `install` or a prefix for such references. And it will get worse, whenever we fully support the distinction of names and paths of submodules. Then, if there was an almighty `install`, such a parameter could be a path to clone from, a name of an existing or not-yet-existing subdataset, or a path to such a subdataset or simply invalid.

So, the idea of using `install` whenever you want to have have something completely unknown to local dataset and `get` otherwise is reasonable to me. And I think it's not hard to get used to it, if you - as a user - are not confused by former notions of `datalad install`.

Does this make it any more clear?

**Comment by yarikoptic** [comment]
this issue was resolved as of 0.4
